lib/eqaf.ml

```diff
@@ -212,3 +220,180 @@ let exists_uint8 ?off ~f str =
+  if digits < 0 then raise (Invalid_argument "digits < 0");
+  let out = Bytes.make digits '0' in
+  let rec loop x = function
+    | -1 -> Bytes.unsafe_to_string out
+    | idx ->
+      let next, this = divmod ~x ~m:10l in
+      Bytes.set out idx @@ char_chr (0x30 lor (Int32.to_int this)) ;
+      loop next (pred idx)
+  in loop n (pred digits)
+
+let[@inline always] to_hex_nibble f : char =
+  let a = 86 + f in
+  let c = 1 + ((a - 71 * ((a land 0x10) lsr 4)) lor 0x20) in
+  char_chr c
+
+let hex_of_string rawbytes =
+  String.init (2 * String.length rawbytes)
+    (fun idx ->
+       let byt = String.get rawbytes (idx lsr 1) |> Char.code in
+       (* select which 4 bits to use, this can probably be done faster:*)
+       let nib = 0xf land (byt lsr (((lnot idx) land 1) lsl 2)) in
+       to_hex_nibble nib)
+
+let hex_of_bytes rawbytes = hex_of_string (Bytes.unsafe_to_string rawbytes)
+
+let[@inline always] lowercase_hex_nibble_to_int ch : int =
+  (* converts [0-9af] to [int] in range [0-15].
+     NB: THIS FUNCTION DOES NOT HANDLE UPPERCASE [A-F] AND DOES NO
+         ERROR DETECTION. ONLY WORKS FOR [0-9a-f] - but then it's fast :-)
+  *)
+  (* camlExample__nibble_to_int_1199:
+        addq    $-32, %rax
+        andq    $-65, %rax
+        movq    %rax, %rbx
+        andq    $45, %rbx
+        shrq    $4, %rbx
+        orq     $1, %rbx
+        imulq   $71, %rbx
+        subq    %rbx, %rax
+        addq    $71, %rax
+        ret
+  *)
+  let x =
+    let x =
+      let i = Char.code ch in
+      (i + 39 - 1 - 54) in
+    x land (lnot 0x20) in
+  x - (((x land 0x16) lsr 4) * 71)
+let _ =
+  (* ended up not using this because error detection was nicer,
+     if a bit slower. *)
+  lowercase_hex_nibble_to_int
+
+let[@inline always] select_a_if_in_range ~low ~high ~n a b =
+  (* select [a] if [low <= n <= high] and [b] if [n] is out of range.*)
+  (* NB: ONLY WORKS FOR [0 <= low <= high <= max_int]*)
+  (* The idea being that:
+     1.a) if low <=   n : (n - low)  is positive +
+     1.b) if low  >   n : (n - low)  is negative -
+     2.a) if   n <= high: (high - n) is positive +
+     2.b) if   n  > high: (high - n) is negative -
+     We OR the numbers together; we only really care about the sign bit
+     which is set when negative.
+     Thus both numbers are positive iff (low <= n && n <= high).
+     We then select the sign bit with (land min_int) and use that to choose:
+  *)
+  let out_of_range = (* choose b if out of range *)
+    ((n - low) lor (high - n)
+     land min_int)
+  in
+  select_int out_of_range a b
+
+let lowercase_ascii src =
+  (* ct version of String.lowercase_ascii *)
+  String.map
+    ( fun ch -> let n = Char.code ch in
+      select_a_if_in_range ~low:0x41 ~high:0x5a ~n (n lor 0x20) (n)
+      |> char_chr
+    ) src
+
+let uppercase_ascii src =
+  (* ct version of String.uppercase_ascii *)
+  String.map
+    ( fun ch -> let n = Char.code ch in
+      select_a_if_in_range ~low:0x61 ~high:0x7a ~n (n lxor 0x20) (n)
```

Comment:
Maybe add comments here about `0x61 = 'a'` and `0x7a = 'z'`, same for lowercase function.
Also that `0x20` controls case for ASCII letters.